core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedBeanFactoryRegistry.java

```diff
@@ -115,13 +118,32 @@ static AnnotatedBeanFactory<?> find(@Nullable BeanFactoryId beanFactoryId) {
 
     static Set<AnnotatedValueResolver> uniqueResolverSet() {
         return new TreeSet<>((o1, o2) -> {
-            final String o1Name = o1.httpElementName();
-            final String o2Name = o2.httpElementName();
-            if (o1Name.equals(o2Name) && o1.annotationType() == o2.annotationType()) {
+            String o1Name = o1.httpElementName();
+            String o2Name = o2.httpElementName();
+            final Class<? extends Annotation> o1AnnotationType = o1.annotationType();
+            final Class<? extends Annotation> o2AnnotationType = o2.annotationType();
+            if (o1AnnotationType == o2AnnotationType && o1Name.equals(o2Name)) {
                 return 0;
             }
-            // We are not ordering, but just finding duplicate elements.
-            return -1;
+
+            // TreeSet internally creates a binary tree. A consistent comparison for the two inputs is
+            // necessary to traverse the binary tree correctly and check uniqueness.
+            // If compare(o1, o2) returns -1, compare(o2, o1) should return 1 or a positive number.
+            if (o1AnnotationType != null) {
+                o1Name += '/' + o1AnnotationType.getName();
+            }
+            if (o2AnnotationType != null) {
+                o2Name += '/' + o2AnnotationType.getName();
+            }
+            final int result = o1Name.compareTo(o2Name);
+            if (result != 0) {
+                return result;
+            }
+            // It is still not a good idea to return 0 since the equality for `httpElementName()` and
+            // `annotationType()` is broken. `o1AnnotationType` and `o2AnnotationType` may be the same
+            // class, but loaded by two class different loaders. The hash comparison is used to return a
+            // non-zero value for the different classes.
+            return Ints.compare(Objects.hashCode(o1AnnotationType), Objects.hashCode(o2AnnotationType));
```

Comment:
I wonder if we could've just removed everything and just returned hashcode comparison (since it would be consistent) 😄   Anyways, I think the current implementation is fine as well 👍 

```suggestion
            // TreeSet internally creates a binary tree. A consistent comparison for the two inputs is
            // necessary to traverse the binary tree correctly and check uniqueness.
            // If compare(o1, o2) returns -1, compare(o2, o1) should return 1 or a positive number.
            return Ints.compare(Objects.hashCode(o1AnnotationType), Objects.hashCode(o2AnnotationType));
```

Reply:
If we exclude `httpElementName()` from comparison, different `httpElementName`s can return 0.
For example, 
```
assert o1.httpElementName() == "name1";
assert o2.httpElementName() == "name2";
assert o1.annotationType() == Param.class;
assert o2.annotationType() == Param.class;
assert o1.annotationType() == o2.annotationType();

assert (o1AnnotationType == o2AnnotationType && o1Name.equals(o2Name)) == false;
assert Ints.compare(Objects.hashCode(o1AnnotationType), Objects.hashCode(o2AnnotationType)) == 0;
```

Reply:
or perhaps we could also include the name when computing the hash as well
```
return Ints.compare(Objects.hashCode(o1.httpElementName(), o1AnnotationType), Objects.hashCode(o2.httpElementName(), o2AnnotationType)) == 0;
```
Having said this, I guess hash functions aren't very deterministic and still run the risk of conflicts, so I think the current implementation is fine 👍

Reply:
Agreed. The two hash functions can return the same value for different values. Thanks for suggesting nice ideas. 🙇‍♂️